source/Variant.cpp

```diff
@@ -0,0 +1,303 @@
+			{
+				reset = false;
+				variants.clear();
+				stockVariants.clear();
+				ships.clear();
+				total = 0;
+				variantTotal = 0;
+				stockTotal = 0;
+			}
+			
+			int n = 1;
+			if(variant)
+			{
+				bool named = false;
+				string variantName;
+				if(child.Size() >= 2 + add && !child.IsNumber(1 + add))
+				{
+					variantName = child.Token(1 + add);
+					named = true;
+					if(variantName == name)
+					{
+						node.PrintTrace("A variant cannot reference itself:");
+						return;
+					}
+				}
+				
+				if(child.Size() >= 2 + add + named && child.Value(1 + add + named) >= 1.)
+					n = child.Value(1 + add + named);
+				total += n;
+				variantTotal += n;
+				
+				// If this variant is named, then look for it in GameData.
+				// Otherwise this is a new variant definition only for this variant.
+				if(named)
+				{
+					stockVariants.emplace_back(GameData::Variants().Get(variantName), n);
+					stockTotal += n;
+				}
+				else
+					variants.emplace_back(child, n);
+			}
+			else
+			{
+				if(child.Size() >= 2 + add && child.Value(1 + add) >= 1.)
+					n = child.Value(1 + add);
+				total += n;
+				ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(0)));
+			}
+		}
+	}
+}
+
+
+
+const string &Variant::Name() const
+{
+	return name;
+}
+
+
+
+vector<const Ship *> Variant::Ships() const
+{
+	return ships;
+}
+
+
+
+vector<pair<const Variant *, int>> Variant::StockVariants() const
+{
+	return stockVariants;
+}
+
+
+
+vector<pair<Variant, int>> Variant::Variants() const
+{
+	return variants;
+}
+
+
+
+// Choose all ships from this variant and choose a ship from each nested variant.
+vector<const Ship *> Variant::ChooseShips() const
+{
+	vector<const Ship *> chosenShips = ships;
+	for(auto &it : variants)
+		for(int i = 0; i < it.second; i++)
+		{
+			vector<const Ship *> variantShips = it.first.NestedChooseShips();
+			chosenShips.insert(chosenShips.end(), variantShips.begin(), variantShips.end());
+		}
+	for(auto &it : stockVariants)
+		for(int i = 0; i < it.second; i++)
+		{
+			vector<const Ship *> variantShips = it.first->NestedChooseShips();
+			chosenShips.insert(chosenShips.end(), variantShips.begin(), variantShips.end());
+		}
+	return chosenShips;
+}
+
+
+
+// Choose a ship from this variant according to the weight of all its contents. 
+vector<const Ship *> Variant::NestedChooseShips() const
+{
+	vector<const Ship *> chosenShips;
+	
+	// Randomly choose between the ships and the variants.
+	int chosen = Random::Int(total);
+	if(chosen < variantTotal)
+	{
+		chosen = Random::Int(variantTotal);
+		unsigned variantIndex = 0;
+		vector<const Ship *> variantShips;
+		// Randomly choose between the stock variants and the non-stock variants.
+		if(chosen < stockTotal)
+		{
+			// Choose a variant according to the weights of the variants.
+			for(int choice = Random::Int(stockTotal); choice >= stockVariants[variantIndex].second; ++variantIndex)
+				choice -= stockVariants[variantIndex].second;
```

Comment:
Some of the code here looks duplicated from the code in Fleet.cpp. Is it possible to use a single/common set of functions instead of 2 sets?

Reply:
There is some duplication, so I'll look in to figuring out if that can be addressed.